### PR TITLE
Implements #94 - Python bindings to get position and rotation based on list of IDs

### PR DESCRIPTION
**Pull request by st4rl3ss**
Implements Python bindings to fetch position and rotations based on list of IDs.
Code has been cleaned up but still requires an ugly and underperforming trick at some point in the code in order to assign/copy the `boost::multiarray`.

### CODE REVIEW

File: python/mvdtool.cpp [hunk 1]
```diff
@@ -167,6 +198,11 @@ PYBIND11_MODULE(mvdtool, mvd) {
                 auto res = f.getPositions(r);
                 return py::array({res.shape()[0], res.shape()[1]}, res.data());
              })
+        .def("positions", [](const File& f, const pyarray<size_t>& idx) {
+                const auto& func = [&f](const MVD::Range& r){return f.getPositions(r);};                
+                const auto res = _atIndicesMulti<Positions, 3>(func, f.size(), idx);
+                return py::array({res.size(), res[0].size()}, res[0].data());
```
Comment (st4rl3ss) [inline python/mvdtool.cpp:204]: Not sure why the `res[0].data` is necessary here but otherwise I would get an higher level double-addressable structure.

**Comment by st4rl3ss** [comment]
please retest

File: python/mvdtool.cpp (hunk 1 above)
Comment (ferdonline) [reply python/mvdtool.cpp]: res is a `vector<array<T>>` so indeed a good way of getting the base address is doing `data()` in the first array of the vector.

File: setup.cfg [hunk 2]
```diff
@@ -2,4 +2,5 @@
 test = pytest
 
 [tool:pytest]
+# addopts = -s  # no stream capture
```
Comment (tristan0x) [inline setup.cfg:5]: dead code?

File: python/mvdtool.cpp [hunk 3]
```diff
@@ -108,31 +108,42 @@ namespace {  // Intenal
...
 
-    for (size_t i=0; i < count; i++) {
-        const size_t elem_i = indices[i];
-        if(elem_i - offset >= chunk.size()) {
-            offset = elem_i;
-            chunk = f(Range(offset, std::min(CHUNK_SIZE, n_records - offset)));
+    for (size_t i=0; i < count;) {
+        const size_t offset = indices[i];
+        const auto limit = std::min(CHUNK_SIZE, n_records - offset);
+        const auto chunk = f(Range(offset, limit));
```
Comment (tristan0x) [inline python/mvdtool.cpp:140]: Not exactly sure what type is `chunk` but I guess you can force copy elusion:
```suggestion
        const auto& chunk = f(Range(offset, limit));
```

**Review by tristan0x — COMMENTED** [review]
Pretty cool! Looks quite efficient!
I have very minor remarks though.

File: python/mvdtool.cpp [hunk 4]
```diff
@@ -155,31 +168,43 @@ PYBIND11_MODULE(mvdtool, mvd) {
...
                 Range r(offset, count);
                 auto res = f.getRotations(r);
-                return py::array({res.shape()[0], res.shape()[1]}, res.data());
+                return py::array({res.shape()[0], ROTATION_WIDTH}, res.data());
+             })
+        .def("rotations", [](const File& f, const pyarray<size_t>& idx) {
+                using rotation_t = std::array<typename Rotations::element, ROTATION_WIDTH>;
+                const auto& func = [&f](const MVD::Range& r){return f.getRotations(r);};
+                const auto res = _atIndices<rotation_t>(func, f.size(), idx);
+                return py::array({res.size(), ROTATION_WIDTH}, res.empty()? nullptr: res[0].data());
```
Comment (tristan0x) [inline python/mvdtool.cpp:207]: Nitpicking here but `front` does not assume `res` is a random access container.
```suggestion
                return py::array({res.size(), ROTATION_WIDTH}, res.empty()? nullptr: res.front().data());
```

File: python/mvdtool.cpp [hunk 5]
```diff
@@ -155,31 +168,43 @@ PYBIND11_MODULE(mvdtool, mvd) {
...
                 Range r(offset, count);
                 auto res = f.getPositions(r);
-                return py::array({res.shape()[0], res.shape()[1]}, res.data());
+                return py::array({res.shape()[0], POSITION_WIDTH}, res.data());
+             })
+        .def("positions", [](const File& f, const pyarray<size_t>& idx) {
+                using position_t = std::array<typename Positions::element, POSITION_WIDTH>;
+                const auto& func = [&f](const MVD::Range& r){return f.getPositions(r);};
+                const auto res = _atIndices<position_t>(func, f.size(), idx);
+                return py::array({res.size(), POSITION_WIDTH}, res.empty()? nullptr: res[0].data());
```
Comment (tristan0x) [inline python/mvdtool.cpp:187]: Nitpicking here but `front` does not assume `res` is a random access container.
```suggestion
                return py::array({res.size(), POSITION_WIDTH}, res.empty()? nullptr: res.front().data());
```

**Comment by ferdonline** [comment]
Thanks man, always nice reviews!

File: python/mvdtool.cpp (hunk 3 above)
Comment (ferdonline) [reply python/mvdtool.cpp]: f is supposed to return a constructed value (RVO), but yep, safer and won't hurt!

File: python/mvdtool.cpp [hunk 6]
```diff
@@ -108,31 +108,42 @@ namespace {  // Intenal
...
+/// Generic routine to copy data item
+template <typename T>
+inline void copy_element(T& dst, const T& src) {
+    dst = src;
+}
+
+/// Copy routine for boost multi_index view to std::array
+template <typename T, size_t Width, typename T2>
+inline void copy_element(std::array<T, Width>& dst, const T2& src) {
+    std::memcpy(&dst[0], src.origin(), sizeof(T) * Width);
```
[the remark this answers is not in the document]
Comment (ferdonline) [reply python/mvdtool.cpp]: The simpler way I can think to SFINAE that out is
```c++
template <typename T, size_t Width, typename T2, typename = decltype(&T2::origin)>
```
@tristan0x what do you think?

**Review by ferdonline — APPROVED** [review]

**Comment by ferdonline** [comment]
@mgeplf Sorry after the merge I didn't follow up much here. That is totally true, I just don't think it's worth to complicate for absolute protection here as it's all internal API. Most type_trait solutions tend to be less readable, but if you have a suggestion please share.

**Comment by mgeplf** [comment]
No worries, depends on how complete you want to be. Readability won't be easy as start as you start adding any checks, I would say.  Not sure if `std::is_function` would do it, but my gut says no.  Prob. need `std::is_member_function_pointer`

**Pull request merged**